src/extension/vectortiles/src/main/java/org/geotools/renderer/lite/VectorMapRenderUtils.java

```diff
@@ -443,4 +445,42 @@ private static boolean isFeatureTypeStyleActive(FeatureType ftype, FeatureTypeSt
                                 || FeatureTypes.isDecendedFrom(
                                         ftype, null, fts.getFeatureTypeName())));
     }
+
+    /**
+     * This function executes transformations on the vector features e.g. point stacker.
+     *
+     * @param layer the layer to process
+     * @param paintArea the area that is painted on
+     * @param mapScale the current map scale
+     * @param schema the schema of the features
+     * @param features the feature collection typically points
+     * @return the transformed feature collection
+     * @throws IOException
+     */
+    public static FeatureCollection<?, ?> transformIfNecessary(
+            Layer layer,
+            Rectangle paintArea,
+            double mapScale,
+            FeatureType schema,
+            FeatureCollection<?, ?> features)
+            throws IOException {
+
+        List<LiteFeatureTypeStyle> featureTypeStyles =
+                VectorMapRenderUtils.getFeatureStyles(layer, paintArea, mapScale, schema);
+        if (!featureTypeStyles.isEmpty()) {
+            for (LiteFeatureTypeStyle featureTypeStyle : featureTypeStyles) {
+                try {
+                    if (featureTypeStyle.transformation != null) {
+                        Expression transform = featureTypeStyle.transformation;
+                        features = (FeatureCollection<?, ?>) transform.evaluate(features);
```

Comment:
This gets reset after each loop; if you have more then one feature type style you will end up only returning features for the last one.

You'll need to combine all the feature collections from each iteration of the loop and return that, rather than just the last one.

Additionally, this does not account for a mix of transformed and not transformed feature collections - at current, only the transformed features will be returned.

It may make more sense to handle transforms as part of the layer handling in VectorTileMapOutputFormat.java and in cases where there are a mix of transformed and not transformed collections treat each as a separate layer


Given the number of different possible configurations involved, I would also recommend a test case for each (No transformation, only a transformation, a transformation and a normal featuretypestyle, an invalid transformation).

Reply:
Hi,
I don't thing that it gets reset. It will just apply the transformations one after another if you do e.g. first a buffer and afterwards a union it will return all the merged polygons.
I have to admit that I did not study the SLD specification on the semantics of how transformations should be applied.

Reply:
Thats a good enough reason to add a test case to verify what actually happens. I'm not certain myself, but the fact that you are reassigning the feature collection each iteration of the loop suggests that something is likely to get overwritten if you have multiple featureTypeStyles